libbpf-tools/offwaketime.c

```diff
@@ -0,0 +1,419 @@
+
+  print_target:
+    printf("    target: %-16s %s (%d)\n", "-", next_key.target.comm,
+           next_key.target.pid);
+    printf("        %lu\n\n", val);
+  }
+
+cleanup:
+  free(ip);
+}
+
+int main(int argc, char **argv) {
+  static const struct argp argp = {
+      .options = opts,
+      .parser = parse_arg,
+      .doc = argp_program_doc,
+  };
+  struct syms_cache *syms_cache = NULL;
+  struct ksyms *ksyms = NULL;
+  const struct ksym *ksym;
+  struct offwaketime_bpf *obj;
+  int err;
+
+  err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+  if (err)
+    return err;
+  if (env.user_threads_only && env.kernel_threads_only) {
+    fprintf(
+        stderr,
+        "user_threads_only and kernel_threads_only cannot be used together.\n");
+    return 1;
+  }
+  if (env.min_block_time >= env.max_block_time) {
+    fprintf(stderr, "min_block_time should be smaller than max_block_time\n");
+    return 1;
+  }
+
+  libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+  libbpf_set_print(libbpf_print_fn);
+
+  obj = offwaketime_bpf__open();
+  if (!obj) {
+    fprintf(stderr, "failed to open BPF object\n");
+    return 1;
+  }
+
+  /* initialize global data (filtering options) */
+  obj->rodata->targ_tgid = env.pid;
+  obj->rodata->targ_pid = env.tid;
+  obj->rodata->user_threads_only = env.user_threads_only;
+  obj->rodata->kernel_threads_only = env.kernel_threads_only;
+  obj->rodata->state = env.state;
+  obj->rodata->min_block_ns = env.min_block_time;
+  obj->rodata->max_block_ns = env.max_block_time;
+
+  bpf_map__set_value_size(obj->maps.stackmap,
+                          env.perf_max_stack_depth * sizeof(unsigned long));
+  bpf_map__set_max_entries(obj->maps.stackmap, env.stack_storage_size);
+
+  err = offwaketime_bpf__load(obj);
+  if (err) {
+    fprintf(stderr, "failed to load BPF programs\n");
+    goto cleanup;
+  }
+  ksyms = ksyms__load();
+  if (!ksyms) {
+    fprintf(stderr, "failed to load kallsyms\n");
+    goto cleanup;
+  }
+  syms_cache = syms_cache__new(0);
+  if (!syms_cache) {
+    fprintf(stderr, "failed to create syms_cache\n");
+    goto cleanup;
+  }
+  ksym = ksyms_get_symbol_match(ksyms, "try_to_wake_up*");
+  if (!ksym) {
+    fprintf(stderr, "failed to find hook point\n");
+    goto cleanup;
+  }
+  obj->links.try_to_wake_up =
+      bpf_program__attach_kprobe(obj->progs.try_to_wake_up, false, ksym->name);
+  ksym = ksyms_get_symbol_match(ksyms, "finish_task_switch*");
+  if (!ksym) {
+    fprintf(stderr, "failed to find hook point\n");
+    goto cleanup;
+  }
+  obj->links.on_cpu =
+      bpf_program__attach_kprobe(obj->progs.on_cpu, false, ksym->name);
+
+  signal(SIGINT, sig_handler);
+
+  /*
+   * We'll get sleep interrupted when someone presses Ctrl-C (which will
+   * be "handled" with noop by sig_handler).
+   */
+  sleep(env.duration);
+
+  print_map(ksyms, syms_cache, obj);
+
+cleanup:
+  offwaketime_bpf__destroy(obj);
+  syms_cache__free(syms_cache);
+  ksyms__free(ksyms);
+  return err != 0;
```

Comment:
```suggestion
  cleanup_core_btf(&open_opts);

  return err != 0;
```

Reply:
@mauriciovasquezbernal  thanks for pointing out. will fix it.